Source/DafnyCore/Backends/Rust/Dafny-compiler-rust.dfy

```diff
@@ -849,16 +974,28 @@ module RAST
           match op2 {
             case "*" | "/" | "%" => PrecedenceAssociativity(20, LeftToRight)
             case "+" | "-" => PrecedenceAssociativity(30, LeftToRight)
-            case "<<" | ">>" => PrecedenceAssociativity(40, LeftToRight)
+            case "<<" | ">>" =>
+              // x as u16 << 6 is parsed as x as u16<... and expect a generic argument
+              if op2 == "<<" && left.TypeAscription? && left.tpe.EndsWithNameThatCanAcceptGenerics() then
```

Comment:
The complexity you mention is due to a unique problem in the Rust lexer, which is that it won't consider "<<" as a single token if it's preceded by an identifier, as it will automatically consider the first "<" as the start of a generic type application, even there is a space between both. So I've seen it complain about not being to parse code, and parentheses would solve the issue.